eval.go

```diff
@@ -568,35 +568,36 @@ func (d *indexData) gatherBranches(docID uint32, mt matchTree, known map[matchTr
 }
 
 func (d *indexData) List(ctx context.Context, q query.Q, opts *ListOptions) (rl *RepoList, err error) {
-	var include func(rle *RepoListEntry) (bool, error)
+	var include func(rle *RepoListEntry) bool
 
 	q = d.simplify(q)
 	if c, ok := q.(*query.Const); ok {
 		if !c.Value {
 			return &RepoList{}, nil
 		}
-		include = func(rle *RepoListEntry) (bool, error) {
-			return true, nil
+		include = func(rle *RepoListEntry) bool {
+			return true
 		}
 	} else {
-		// We need to run a search per repo to decide if it is included.
-		include = func(rle *RepoListEntry) (bool, error) {
-			qOneRepo := query.NewAnd(
-				query.NewRepoSet(rle.Repository.Name),
-				q)
-			sr, err := d.Search(ctx, qOneRepo, &SearchOptions{
-				ShardMaxMatchCount: 1,
-				TotalMaxMatchCount: 1,
-			})
-			if err != nil {
-				return false, err
-			}
-			return len(sr.Files) > 0, nil
+		sr, err := d.Search(ctx, q, &SearchOptions{
+			ShardRepoMaxMatchCount: 1,
+		})
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Instead of running a search per repo, this changes it to run a single ahead-of-time search that limits the results to a single match per repo. This should be much more efficient for a large number of repos.